src/encoded/static/components/top_hits/hit.js

```diff
@@ -0,0 +1,174 @@
+
+Title.propTypes = { value: PropTypes.string.isRequired };
+
+Other.propTypes = { value: PropTypes.string.isRequired };
+
+
+/**
+* Maps a field to the formatting component used to style
+* that part of the hit.
+*/
+const fieldToComponent = {
+    name: Accession,
+    title: Title,
+    target: Other,
+    details: Other,
+    description: Description,
+};
+
+
+/**
+* Shortens a sentence to maxWords.
+*/
+export const maybeTrimByWord = (value, maxWords) => {
+    const splitValue = value.split(WHITESPACE);
+    if (splitValue.length > maxWords) {
+        return `${splitValue.slice(0, maxWords).join(WHITESPACE)}...`;
+    }
+    return value;
+};
+
+
+/**
+* Some fields are lists and should be converted to strings.
+*/
+export const maybeConvertArrayToString = (value, sep = ', ') => {
+    if (Array.isArray(value)) {
+        return value.join(sep);
+    }
+    return value;
+};
+
+
+/**
+* Encapsulates all the formatting for a raw top hit result. This
+* tries to generalize across most objects and return the first
+* value for that category that exists. It's important to remember
+* that these fields have to be specifically requested in the params
+* of the query URL (i.e. ?field=param) defined in ./constants.
+*/
+class Hit {
+    constructor(item) {
+        this.item = item;
+    }
+
+    formatTitle() {
+        return (
+            this.item.assay_title ||
+            this.item.assay_term_name ||
+            this.item.term_name
+        );
+    }
+
+    formatTarget() {
+        return this.item.target && this.item.target.title;
+    }
+
+    formatName() {
+        return this.item.accession || this.item['@id'];
+    }
+
+    formatDescription() {
+        return (
+            this.item.biosample_summary ||
+            this.item.summary ||
+            this.item.description ||
+            this.item.output_type ||
+            this.item.classification ||
+            this.item.antigen_description ||
+            this.item.title
+        );
+    }
+
+    formatDetails() {
+        return this.item.file_type;
+    }
+
+    /**
+    * This controls the order of values that are displayed in a
+    * result. The field string is used to recover the appropriate
+    * formatting component and also as a unique rendering key. The
+    * values are converted to a string (in case they are a list), and
+    * then empty values are filtered out (e.g. not every object
+    * has a target etc.).
+    */
+    getValues() {
+        return [
+            ['name', this.formatName()],
+            ['title', this.formatTitle()],
+            ['target', this.formatTarget()],
+            ['details', this.formatDetails()],
+            ['description', this.formatDescription()],
+        ].map(
+            ([field, value]) => (
+                [
+                    field,
+                    maybeConvertArrayToString(value),
+                ]
+            )
+        ).filter(
+            ({ 1: value }) => Boolean(value)
+        );
+    }
```

Comment:
I think there are a couple disadvantages for calling it in the list. One is that it will require duplicating the call for every item, and any future items that are added. The second is that it will make it less flexible if there is ever an intermediate formatting step that needs to be inserted between getting the values and converting them to arrays.

Reply:
The duplication is more of a copy/paste cost though. The cost for execution or memory would be negligible since maybeConvertArrayToString isn’t inline. For inserting a new step, yeah, doing it all in one means more copy/paste, where doing it in a separate map keeps any future changes in one place. I’m fine either way. The way you have it involves another loop and another inline function which have their own costs, but probably close to negligible as well.

Reply:
Yeah, maybe a bit of opinion. If it was one or two items in a list that was never going to change I think your way is clearer. Since it's possible we'll want to parse other values in object types, and they might require additional formatting steps, I think this violates DRY less and makes it clearer that a uniform step is applied to all values. I think worst case you can imagine the visual clutter cost of unrolling something like: 

```
const results = [
    ['name', this.formatName()],
    ['title', this.formatTitle()],
    ['target', this.formatTarget()],
    ['details', this.formatDetails()],
    ['description', this.formatDescription()],
].map(
    ([field, value]) => (
        [
            field,
            maybeDoSomethingElseAfter(
                maybeConvertArrayToString(
                    maybeDoSomethingElseFirst(
                        value
                    ),
                ),
            ),
        ]
    )
);
```
to 
```
const results = [
    [
        'name',
        maybeDoSomethingElseAfter(
            maybeConvertArrayToString(
                maybeDoSomethingElseFirst(
                    this.formatName()
                ),
            ),
        ),
    ],
    [
        'title',
        maybeDoSomethingElseAfter(
            maybeConvertArrayToString(
                maybeDoSomethingElseFirst(
                    this.formatTitle()
                ),
            ),
        ),
    ],
    [
        'target',
        maybeDoSomethingElseAfter(
            maybeConvertArrayToString(
                maybeDoSomethingElseFirst(
                    this.formatTarget()
                ),
            ),
        ),
    ],
    [
        'details',
        maybeDoSomethingElseAfter(
            maybeConvertArrayToString(
                maybeDoSomethingElseFirst(
                    this.formatDetails()
                ),
            ),
        ),
    ],
    [
        'description',
        maybeDoSomethingElseAfter(
            maybeConvertArrayToString(
                maybeDoSomethingElseFirst(
                    this.formatDescription()
                ),
            ),
        ),
    ],
];
```
I think it's harder to determine if all the values go through the same steps in the second example, and also a pain if you change the name of one of the functions. I don't think the immutable functional folks would care much about the cost of creating a new list.

Reply:
I would definitely prefer your way if we needed anything remotely like that.